tests/assoc_sim.R

```diff
@@ -0,0 +1,143 @@
+# Inputs: N = number of reports
+#         candidates = list containing a list of candidate strings
+#         params = list with RAPPOR parameters
+#         mapfile = file to write maps into (with .csv suffixes)
+#         reportsfile = file to write reports into (with .csv suffix)
+#         unif = whether to replace poisson with uniform
+SimulateReports <- function(N, candidates, params, mapfile, reportsfile,
+                            unif) {
+  # Compute true distribution
+  m <- params$m  
+  samples <- list()
+
+  if (unif) {
+    # Draw uniformly from 1 to 10
+    samples[[1]] <- as.integer(runif(N, 1, 10))
+  } else {
+    # Draw from a Poisson random variable
+    samples[[1]] <- rpois(N, 1) + rep(1, N)
+  }
+  
+  # Pr[var2 = N + 1 | var1 = N] = 0.5
+  # Pr[var2 = N     | var1 = N] = 0.5
+  samples[[2]] <- samples[[1]] + sample(c(0, 1), N, replace = TRUE)
+  
+  # Replace sample i with candidate i for each variable
+  for(i in 1:2) {
+    if(max(samples[[i]]) <= length(candidates[[i]])) {
+      samples[[i]] <- candidates[[i]][samples[[i]]]
+    } else {
+      # Pad candidates with sample strings
+      # If only 2 candidates, new set of candidates becomes:
+      # candidate1, candidate2, str3, str4, ...
+      len <- length(candidates)
+      candidates[[i]][(len + 1):max(samples[[i]])] <- apply(
+        as.matrix((len + 1):max(samples[[i]])),
+        1,
+        function(i) paste("str", as.character(i), sep = ""))
+      samples[[i]] <- candidates[[i]][samples[[i]]]
+    }
+  }
+  
+  # Randomly assign cohorts in each dimension
+  cohorts <- sample(1:m, N, replace = TRUE)
+  
+  # Create and write map into mapfile_1.csv and mapfile_2.csv
+  map <- lapply(1:2, function(i) CreateMap(candidates[[i]], params))
+  write.table(map[[1]]$map_pos, file = paste(mapfile, "_1.csv", sep = ""),
+              sep = ",", col.names = FALSE, na = "", quote = FALSE)
+  write.table(map[[2]]$map_pos, file = paste(mapfile, "_2.csv", sep = ""),
+              sep = ",", col.names = FALSE, na = "", quote = FALSE)
+  
+  # Write reports into reportsfile.csv
+  # Format:
+  #     cohort, bloom filter var1, bloom filter var2
+  reports <- lapply(1:2, function(i)
+    EncodeAll(samples[[i]], cohorts, map[[i]]$map, params))
+  # Organize cohorts and reports into format
+  write_matrix <- cbind(as.matrix(cohorts),
+                        as.matrix(lapply(reports[[1]], 
+                            function(x) paste(x, collapse = ""))),
+                        as.matrix(lapply(reports[[2]],
+                            function(x) paste(x, collapse = ""))))
+  write.table(write_matrix, file = reportsfile, quote = FALSE,
+              row.names = FALSE, col.names = FALSE, sep = ",")
+}
+
+main <- function(opts) {
+  ptm <- proc.time()
+  
+  candidates <- GetCandidatesFromFile(opts$candidates)
+  params <- ReadParameterFile(opts$params)
+  SimulateReports(opts$num, candidates, params, opts$map, opts$reports, opts$unif)
```

Comment:
I like to structure arguments so all the inputs are first and the outputs are last.  Here the outputs opts$map and opts$reports are sandwiched between inputs.